CraftBukkit/0161-Add-PlayerSettingsUpdateEvent.patch

```diff
@@ -0,0 +1,298 @@
+     }
+ 
++    // SportBukkit start - player settings
++    @Override
++    public PlayerSettings getSettings() {
++        return getHandle().settings;
++    }
++
++    @Override
++    public void setSettings(PlayerSettings settings) {
++        getHandle().settings = settings;
++    }
++
+     @Override
+     public String getLocale() {
+-        final String locale = getHandle().locale;
+-        return locale != null ? locale : "en_US";
++        return getSettings().getLocale();
+     }
+ 
+     @Override
+     public void setLocale(String locale) {
+-        getHandle().locale = locale;
++        setSettings(new PlayerSettings.Delta(getSettings(), locale, null, null, null, null, null));
+     }
++    // SportBukkit end
+ 
+     @Override
+     public java.util.Locale getCurrentLocale() {
+diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+index 1f6916f..a3cc07a 100644
+--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
++++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+@@ -6,6 +6,7 @@ import java.util.ArrayList;
+ import java.util.EnumMap;
+ import java.util.List;
+ import java.util.Map;
++import java.util.Set;
+ 
+ import com.google.common.base.Function;
+ import com.google.common.base.Functions;
+@@ -13,11 +14,14 @@ import com.google.common.base.Functions;
+ import net.minecraft.server.*;
+ 
+ import org.bukkit.Bukkit;
++import org.bukkit.ChatVisibility;
+ import org.bukkit.EntityLocation;
+ import org.bukkit.Location;
+ import org.bukkit.Material;
++import org.bukkit.PlayerSettings;
+ import org.bukkit.PoseFlag;
+ import org.bukkit.Server;
++import org.bukkit.Skin;
+ import org.bukkit.Statistic.Type;
+ import org.bukkit.block.Block;
+ import org.bukkit.block.BlockFace;
+@@ -35,6 +39,7 @@ import org.bukkit.craftbukkit.inventory.CraftItemStack;
+ import org.bukkit.craftbukkit.inventory.CraftMetaBook;
+ import org.bukkit.craftbukkit.util.CraftDamageSource;
+ import org.bukkit.craftbukkit.util.CraftMagicNumbers;
++import org.bukkit.craftbukkit.util.Skins;
+ import org.bukkit.entity.AreaEffectCloud;
+ import org.bukkit.entity.Arrow;
+ import org.bukkit.entity.Creeper;
+@@ -65,7 +70,10 @@ import org.bukkit.event.player.*;
+ import org.bukkit.event.server.ServerListPingEvent;
+ import org.bukkit.inventory.EquipmentSlot;
+ import org.bukkit.inventory.InventoryView;
++import org.bukkit.inventory.MainHand;
+ import org.bukkit.inventory.meta.BookMeta;
++import org.bukkit.plugin.PluginManager;
++import org.bukkit.util.EnumUtils;
+ 
+ public class CraftEventFactory {
+     public static final DamageSource MELTING = CraftDamageSource.copyOf(DamageSource.BURN);
+@@ -1053,11 +1061,23 @@ public class CraftEventFactory {
+         return event;
+     }
+ 
+-    public static PlayerLocaleChangeEvent callPlayerLocaleChangeEvent(EntityHuman who, String oldLocale, String newLocale) {
```

Comment:
If this patch is overwriting a lot of stuff from the locale change event patch, then it should just be combined with that patch. When we rebase, we don't want to merge code that gets thrown away in later patches.